### PR TITLE
[ONNX] Create decorator to handle symbolic context

### LINKED ISSUE #84553: [ONNX] Change how context is given to symbolic functions
Current symbolic functions can take a context as an input, pushing graphs to the second argument. To support these functions, we need to annotate the first argument as symbolic context and tell them part in call time by examining the annotations. 

Checking annotations is slow and this process complicates the logic in the caller. 

Instead we can wrap the graph object in a GraphContext, exposing all methods used from the graph and include the context in the GraphContext. This way all the old symbolic functions continue to work and we do not need to do the annotation checking if we know the symbolic function is a "new function". 

We can edit a private field in the functions at registration time to tag them as "new style" symbolic functions that always takes a wrapped Graph with context object as input.

This also has the added benefit where we no longer need to monkey patch the Graph object to expose the g.op method. Instead the method can be defined in the graph context object.

### PR DESCRIPTION
**Pull request by justinchuby**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85488
* #85487
* #85486
* #85350
* #85349
* #85312
* #85165
* #85491
* __->__ #84776
* #84728

- Create decorator to handle old style custom symbolics that require context
- Deprecate `torch.onnx.SymbolicContext` in favor of `GraphContext`. Added deprecation message
- Remove README reference of SymbolicContext

### CODE REVIEW

File: torch/onnx/_exporter_states.py [hunk 1]
```diff
@@ -25,6 +26,12 @@ class SymbolicContext:
         onnx_block (_C.Block): Current ONNX block that converted nodes are being appended to.
     """
 
+    @_deprecation.deprecated(
+        "1.13",
+        "TBD",
+        # TODO(justinchuby): Fix the instruction when GraphContext is public.
+        "remove the 'ctx' argument and annotate 'g: GraphContext' instead",
```
Comment (AllenTiTaiWang) [inline torch/onnx/_exporter_states.py:33]: I thought this was still under discusstion😂

File: torch/onnx/_exporter_states.py [hunk 2]
```diff
@@ -25,6 +26,12 @@ class SymbolicContext:
...
+    )
```
Comment (AllenTiTaiWang) [inline torch/onnx/_exporter_states.py:34]: Will `SymbolicContext` be deprecated?

**Review by titaiwangms — APPROVED** [review]
LGTM!

File: torch/onnx/_exporter_states.py (hunk 2 above)
Comment (justinchuby) [reply torch/onnx/_exporter_states.py:34]: I intended to deprecate SymbolicContext. @BowenBao @shubhambhokare1 @thiagocrepaldi any thoughts?

**Review by thiagocrepaldi — COMMENTED** [review]
@BowenBao and I were discussing a similar issue (passing context to some symbolics, but in a deeper level  - that we can discuss separately to prevent noise here).

In short, this context would *also* be populated by the JIT [script?] C++ core, with `dtype`s, `shape` and whatever is available so that it can assist symbolic implementors during `torch.onnx.export` for models with `torch.jit.script` layers. For scripted layers, hard-to-guess construct happen, such as `Tensor[]` in which we cannot know whether it was dtype=float or a list of something else, for example (that seems to be avaiable during torch.jit.script time).

Regarding to this PR in specific, I wonder if we could *always* make the existing context available to all symbolics by default, making it a generic solution for any new symbolic/opset to come. 

For this approach, an idea would be tapping the context into [Graph](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/ir/ir.h#L1165) and/or adding a [python binding method to Graph](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/python/python_ir.cpp#L218)

Although this is a change to the JIT IR core, if we do it in a nice way (low coupling and high cohesion), meta might accept changes to `tyorch._C.Graph`. 

With this, we might reuse it not only to solve the current problem (some symbolics need params_dict, env, cur_node, onnx_block), but also will have a standard way to extend `torch.jit.script` support, by providing means to the core to push context there too

@BowenBao what do you think?

**Comment by justinchuby** [comment]
> Regarding to this PR in specific, I wonder if we could *always* make the existing context available to all symbolics by default, making it a generic solution for any new symbolic/opset to come. 

> tapping the context into Graph 

Great idea! Please see #84728 (previous PR on the stack) that implements this. Can you check the discussion section in the PR and share what you think? The decorator exists to support the legacy style (for bc) that specifies an additional context. We expect all new symbolic functions assume the context in `g`.

**Comment by thiagocrepaldi** [comment]
> > Regarding to this PR in specific, I wonder if we could _always_ make the existing context available to all symbolics by default, making it a generic solution for any new symbolic/opset to come.
> 
> > tapping the context into Graph
> 
> Great idea! Please see #84728 (previous PR on the stack) that implements this. Can you check the discussion section in the PR and share what you think? The decorator exists to support the legacy style (for bc) that specifies an additional context. We expect all new symbolic functions assume the context in `g`.

Add a deprecation message to the decorator to let users know about the new mechanism. This way we can try to remove the decorator in a couple releases.

File: torch/onnx/_exporter_states.py (hunk 2 above)
Comment (BowenBao) [reply torch/onnx/_exporter_states.py:34]: Sounds good, since all previous functionalities can be achieved through new GraphContext.

**Review by BowenBao — APPROVED** [review]
LGTM, please add a few test cases for coverage.

File: torch/onnx/_exporter_states.py [hunk 3]
```diff
@@ -25,6 +26,12 @@ class SymbolicContext:
         onnx_block (_C.Block): Current ONNX block that converted nodes are being appended to.
     """
 
+    @_deprecation.deprecated(
+        "1.13",
+        "TBD",
```
Comment (BowenBao) [inline torch/onnx/_exporter_states.py:31]: Forgot to mention, do you intend to keep TBD?

**Review by BowenBao — CHANGES_REQUESTED** [review]
Just noticed that SymbolicContext is also mentioned in our onnx doc page, please update the information there.

File: torch/onnx/_exporter_states.py (hunk 3 above)
Comment (justinchuby) [reply torch/onnx/_exporter_states.py]: Updated to 1.14

**Comment by justinchuby** [comment]
@BowenBao Do you think this test is good enough?

https://github.com/pytorch/pytorch/blob/848437590f41af5d3c9f9bb381106114f70fe572/test/onnx/test_custom_ops.py#L111-L129

File: torch/onnx/_exporter_states.py (hunk 3 above)
Comment (justinchuby) [reply torch/onnx/_exporter_states.py]: README updated. Public user doc to be updated in another PR.

File: torch/onnx/utils.py [hunk 4]
```diff
@@ -45,6 +45,7 @@
     symbolic_helper,
 )
 from torch.onnx._globals import GLOBALS
+from torch.onnx import _deprecation
```
Comment (justinchuby) [inline torch/onnx/utils.py:48]: Remove

File: torch/onnx/utils.py [hunk 5]
```diff
@@ -1776,12 +1802,6 @@ def _run_symbolic_function(
...
                 }
-                if _need_symbolic_context(symbolic_fn):
-                    # TODO(justinchuby): Refactor how we check for the need of the symbolic context
-                    ctx = _exporter_states.SymbolicContext(
-                        _params_dict, env, node, block
-                    )
-                    return symbolic_fn(ctx, graph_context, *inputs, **attrs)
                 # PythonOp symbolic need access to the node to resolve the name conflict,
                 # this is inconsistent with regular op symbolic.
                 if op_name == "PythonOp":
```
Comment (justinchuby) [inline torch/onnx/utils.py:1807]: What happens if a PythonOp needs symbolic_context? @BowenBao how can we handle python ops elegantly?

Comment (justinchuby) [reply torch/onnx/utils.py:1807]: We can remove this.

File: torch/onnx/utils.py [hunk 6]
```diff
@@ -1707,6 +1708,32 @@ def _need_symbolic_context(symbolic_fn) -> bool:
...
 
+@_beartype.beartype
+def _symbolic_context_handler(symbolic_fn: Callable) -> Callable:
+    """Decorator that provides the symbolic context to the symbolic function if needed."""
+    if _need_symbolic_context(symbolic_fn):
+
+        # TODO(justinchuby): Update the module name of GraphContext when it is public
+        warnings.warn(
+            "The first argument to symbolic functions is deprecated in 1.13 and will be "
+            "removed in 1.14. Please annotate the first argument as GraphContext "
```
Comment (justinchuby) [inline torch/onnx/utils.py:1719]: change wording

File: torch/onnx/utils.py (hunk 5 above)
Comment (justinchuby) [reply torch/onnx/utils.py:1808]: Removed

File: torch/onnx/utils.py (hunk 6 above)
Comment (justinchuby) [reply torch/onnx/utils.py:1719]: Done

**Comment by justinchuby** [comment]
@BowenBao PTAL

**Review by BowenBao — APPROVED** [review]
🚀

**Pull request closed without merging**